Fix parsing error of test data

RonEld opened the pull request:
## Description
Fix parsing error that contains special character.
The previous implementation replaced the `:` char with `\n`,
and split on `\n`. Test data containing strings with `\n`
were split as well. Resolves #2193.
The split function caused strings containing `\:` to add
another escape char, resulting in `\\:`. This caused the
tests with the `\:` in the string data to fail.
The fix doesn't replace with `\n`, but splits all `:` that
are not preceded with `\`. After that, removes the preceding `\` char.

## Status
**READY**

## Requires Backporting
NO  

## Additional comments
Any additional information that could be of interest

## Todos
- [x] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported


## Steps to test or reproduce
Outline the steps to test or reproduce the PR here.

Patater commented on tests/scripts/mbedtls_test.py:
```diff
@@ -112,8 +110,8 @@ def __parse(self, data_f):
-                raise TestDataParserError("Number of test arguments should "
-                                          "be even: %s" % line)
+                raise TestDataParserError(("Number of test arguments({}) should "
+                                          "be even: {}").format( args_count, line ) )
```
Does this pass `check-python-files.sh`? I'm curious because the spaces around the `(` are Mbed TLS style and not Python style.

RonEld replied:
Since it's not part of the CI, or part of the pre-push script it wasn't run... but the python itself doesn't crash
The script fails on a pylint error:
`pylint: error: no such option: -j`

RonEld replied:
The pylint version I am using is 1.1.0

Patater replied:
It's part of the CI that gatekeepers have to run manually, also known as `all.sh`. :)

I have `-j` on my version, which is pylint 1.8.3. It just means "run multiple jobs in parallel" and you could omit it. It's present just to speed up all the checks.

RonEld replied:
my `all.sh` never failed. and I had to install pylint only now

After removing the `-j` in the script, I still get pylint errors, investigating at the moment

Patater replied:
Unfortunately, `all.sh` checks for the existence of pylint, and only if it is present does it do the checks. I think that's a bug, as it means we might think we tested for something when we didn't.

RonEld replied:
I create #2252 to run the script when pushing a PR

Patater commented on tests/scripts/mbedtls_test.py:
```diff
@@ -77,9 +77,7 @@ def __escaped_split(inp_str, split_char):
-        out = re.sub(r'(\\.)|' + split_char,
-                     lambda m: m.group(1) or '\n', inp_str,
-                     len(inp_str)).split('\n')
+        out = map(lambda x: re.sub(r'\\'+split_char, split_char,x), re.split(r'(?<!\\)'+split_char, inp_str))
```
A few spaces missing
```diff
- out = map(lambda x: re.sub(r'\\'+split_char, split_char,x), re.split(r'(?<!\\)'+split_char, inp_str))
+ out = map(lambda x: re.sub(r'\\' + split_char, split_char, x), re.split(r'(?<!\\)' + split_char, inp_str))
```

RonEld commented:
@Patater  I addressed your comments, although unfortunately, I still have pylint issues, so I couldn't run the `check-python` script

Patater commented:
@RonEld Are the pylint issues pre-existing?

andresag01 commented on tests/scripts/mbedtls_test.py:
```diff
@@ -77,9 +77,7 @@ def __escaped_split(inp_str, split_char):
-                     lambda m: m.group(1) or '\n', inp_str,
-                     len(inp_str)).split('\n')
+        out = map(lambda x: re.sub(r'\\' + split_char, split_char,x), re.split(r'(?<!\\)' + split_char, inp_str))
         out = [x for x in out if x]
```
I am happy with the new `map()` expression, but I was wondering: Is the line `out = [x for x in out if x]` necessary? It seems we are filtering out some test parameters that evaluate to false (for example `None`?) Why would that happen?

andresag01 commented on tests/scripts/mbedtls_test.py:
```diff
@@ -77,9 +77,7 @@ def __escaped_split(inp_str, split_char):
-        out = re.sub(r'(\\.)|' + split_char,
-                     lambda m: m.group(1) or '\n', inp_str,
-                     len(inp_str)).split('\n')
+        out = map(lambda x: re.sub(r'\\' + split_char, split_char,x), re.split(r'(?<!\\)' + split_char, inp_str))
```
Style: Could you please split up this line to make it shorter?

RonEld commented:
@andresag01 Thanks for your suggestions!
For some reason when I ran the pylint(version 1.9.3) yesterday, on my machine I didn't get those messages, but today I got the same errors you posted.
I fixed according to your suggestions.

@Patater All the issues raised by pylint are not preexisting, and happened with this change.

Patater commented:
Recommend squashing commits into 1, since the later commits just fix issues with the first commit.

RonEld commented:
@Patater Should I rebase now, or should we wait until it is approved by @andresag01 ?

Patater commented:
Up to @andresag01

andresag01 commented:
@RonEld: Its OK from my side if you want to rebase/squash the PR, there is not much history anyways.

andresag01 commented on tests/scripts/mbedtls_test.py:
```diff
@@ -75,12 +75,10 @@ def __escaped_split(inp_str, split_char):
-        out = re.sub(r'(\\.)|' + split_char,
-                     lambda m: m.group(1) or '\n', inp_str,
-                     len(inp_str)).split('\n')
-        out = [x for x in out if x]
```
Are we sure that this line is not needed?

RonEld replied:
Honestly, I don't know. I am not a python expert, and I can't say why it was put here.
This line was accidentally removed... I'll check tomorow

andresag01 replied:
This line of code looks a bit suspicious to me, it is effectively removing all the items in the out list that evaluate to false. I think that the values in the out list can only be strings, so I think this would only remove empty strings from the out list. However, I dont see what situations would produce these empty strings...

RonEld replied:
I don't know what could evaluate to none in this case, but I'm pretty sure it's not the empty strings, as out test data contains empty strings.
I have tried the following:
~~~
>>> inp_str1='exp:2:char*:""'
>>> print inp_str1
exp:2:char*:""
>>> out=map(lambda x: re.sub(r'\\'+split_char, split_char,x), re.split(r'(?<!\\)'+split_char, inp_str1))
>>> print out
['exp', '2', 'char*', '""']
>>> out = [x for x in out if x]
>>> print out
['exp', '2', 'char*', '""']
~~~
As you can see, the empty strings are not removed. Perhaps it's just corner case checking. note that once removing a "none" variable, it is likely that the script will raise an error on odd argument count.

I will return this line, even though I am not sure what it does,

andresag01 replied:
> As you can see, the empty strings are not removed.

I do not think there are any empty strings in your example. The string `'""'` is not empty as it contains the " characters

RonEld replied:
Nevermind, I tried the following:
~~~
>>> inp_str1='exp:2:char*:'
>>> out=map(lambda x: re.sub(r'\\'+split_char, split_char,x), re.split(r'(?<!\\)'+split_char, inp_str1))
>>> print out
['exp', '2', 'char*', '']
>>> out = [x for x in out if x]
>>> print out
['exp', '2', 'char*']
~~~

andresag01 replied:
ok, thanks for the info. I did not know that we allowed terminating the test line with a :

RonEld replied:
we don't allow this, but one can make mistakes.
This line was a measure of checking precondition exists.
also, I tried the following:
~~~
>>> inp_str1='exp::2::char*:'
>>> out=map(lambda x: re.sub(r'\\'+split_char, split_char,x), re.split(r'(?<!\\)'+split_char, inp_str1))
>>> print out
['exp', '', '2', '', 'char*', '']
>>> out = [x for x in out if x]
>>> print out
['exp', '2', 'char*']
~~~
So it's not only a matter of terminating the data with `:` , it's checking the `:` separator is as expected.
I'll return the line

RonEld commented:
@andresag01 I squashed the previous 3 commits, and added a new commit that returns the removed line.

Please review.

once you approve, I will squash the remaining two commits
cc @Patater

RonEld commented:
@andresag01 @Patater I squashed the two commits into a single commit.
Please approve

RonEld commented:
I just noticed a typo in the commit title ( double "of").
I amended the commit
